### ISSUE TITLE
NullPointerException in RoleResolveUtil when admin-cli uses lightweight token

### ISSUE THREAD

**Issue by extbogzav** [keycloak/keycloak#34444]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

admin/cli

### Describe the bug

A NullPointerException is thrown in RoleResolveUtil#addToToken, but I'm not sure how to reproduce it yet, as it only happens sporadically.

```
java.lang.NullPointerException: Cannot invoke "org.keycloak.models.ClientModel.getClientId()" because "app" is null
  at org.keycloak.utils.RoleResolveUtil.addToToken(RoleResolveUtil.java:125)
  at org.keycloak.utils.RoleResolveUtil.lambda$getAndCacheResolvedRoles$0(RoleResolveUtil.java:105)
  at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(Unknown Source)
  at java.base/java.util.stream.ReferencePipeline$Head.forEach(Unknown Source)
  at org.keycloak.utils.RoleResolveUtil.getAndCacheResolvedRoles(RoleResolveUtil.java:105)
  at org.keycloak.utils.RoleResolveUtil.getResolvedRealmRoles(RoleResolveUtil.java:51)
  at org.keycloak.services.resources.admin.permissions.MgmtPermissions.initIdentity(MgmtPermissions.java:126)
  at org.keycloak.services.resources.admin.permissions.MgmtPermissions.<init>(MgmtPermissions.java:100)
  at org.keycloak.services.resources.admin.permissions.AdminPermissions.evaluator(AdminPermissions.java:39)
  at org.keycloak.services.resources.admin.RealmsAdminResource.getRealmAdmin(RealmsAdminResource.java:207)
  at org.keycloak.services.resources.admin.RealmsAdminResource$quarkusrestinvoker$getRealmAdmin_a79304c5f6b93bb8957395d5fd7a2ca5d97a5ae2.invoke(Unknown Source)
  at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
  at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
  at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
  at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:635)
  at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
  at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
  at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
  at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
  at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
  at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  at java.base/java.lang.Thread.run(Unknown Source)
```

To me, it seems it is also related to the admin-cli using a lightweight token because the only way I was able to get rid of this NPE was by reverting back to the old implementation before the lightweight access token was configured for the admin-cli.
The change I tried (which is obviously not a good fix), in `org.keycloak.services.resources.admin.permissions.MgmtPermissions#initIdentity`

```
if (Constants.ADMIN_CLI_CLIENT_ID.equals(issuedFor) && Profile.isFeatureEnabled(Profile.Feature.DYNAMIC_SCOPES)) {
  this.identity = new UserModelIdentity(auth.getRealm(), auth.getUser());
}
```

### Version

26.0.2

### Regression

- [X] The issue is a regression

### Expected behavior

No NullPointerException is thrown

### Actual behavior

NullPointerException is thrown, but it is unclear how to reproduce this reliably.

### How to Reproduce?

Don't know yet.

**Comment by shawkins** [comment]
Adding in core-shared. The thing that looks the most suspect in the immediate code is if a RoleModel from the ClientSessionContext role stream has a client id that is no longer valid such that getContainer returns null.

cc @ahus1 does that seem related to lightweight tokens?

**Comment by ahus1** [comment]
@shawkins - AFAIK I haven't been involved with lightweight tokens, so I'm not sure how I can help here. Let me know if you think otherwise.

**Comment by shawkins** [comment]
> @shawkins - AFAIK I haven't been involved with lightweight tokens, so I'm not sure how I can help here. Let me know if you think otherwise.

Sorry, wasn't sure who to tag on this. @pedroigor or @stianst any ideas?

**Comment by mposolda** [comment]
@graziang Could you triage this please?

**Comment by graziang** [comment]
Lightweight access token for the admin REST API uses `RoleResolveUtil.getAllResolvedClientRoles()` to resolve roles, as they are missing in the token. But it seems the issue isn’t really about lightweight tokens directly, it's more about resolving roles.

The stack traces shared by @extbogzav and @mkopeyka are a bit different, but both show an error at `org.keycloak.utils.RoleResolveUtil.addToToken(RoleResolveUtil.java:125)`. This happens because `getContainer()` which calls `realm.getClientById()` is returning null. This is probably due to a missing client or an incorrect ID parameter, as @shawkins pointed out.

@mkopeyka, does this issue happen every time for you? If so, could you please share more details about the client configuration you're using?

**Comment by dimtoneff** [comment]
Upgraded from v23 to v26.

Things seems to be working but I can't login to the admin console:


`2024-11-15 13:53:48,342 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-10) Uncaught server error: java.lang.NullPointerException: Cannot invoke "org.keycloak.models.ClientModel.getClientId()" because "app" is null
	at org.keycloak.utils.RoleResolveUtil.addToToken(RoleResolveUtil.java:125)
	at org.keycloak.utils.RoleResolveUtil.lambda$getAndCacheResolvedRoles$0(RoleResolveUtil.java:105)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1715)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at org.keycloak.utils.RoleResolveUtil.getAndCacheResolvedRoles(RoleResolveUtil.java:105)
	at org.keycloak.utils.RoleResolveUtil.getResolvedRealmRoles(RoleResolveUtil.java:51)
	at org.keycloak.services.resources.admin.permissions.MgmtPermissions.initIdentity(MgmtPermissions.java:126)
	at org.keycloak.services.resources.admin.permissions.MgmtPermissions.<init>(MgmtPermissions.java:107)
	at org.keycloak.services.resources.admin.permissions.AdminPermissions.realms(AdminPermissions.java:46)
	at org.keycloak.services.resources.admin.AdminRoot.getServerInfo(AdminRoot.java:263)
	at org.keycloak.services.resources.admin.AdminRoot$quarkusrestinvoker$getServerInfo_af58ef46002deb1ad4721c04a79203aeec92de8e.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:635)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
`

**Comment by dimtoneff** [comment]
> It’s possible that a client role is somehow linked to a client that no longer exists. Could you please share a realm export, or at least just the section related to clients, client roles and client scopes?

@graziang 
Unfortunately I can not get a permission to share any Database dumps, but I was able to fix the issue with the data.

You were basically right. There were 18 roles from KEYCLOAK_ROLE which were referring to a CLIENT which was no longer existing. This KC installation is quite old and was upgraded multiple times.

The following query shows the culprit entries:

> SELECT kr.id, kr.NAME, kr.`CLIENT`, kr.CLIENT_REALM_CONSTRAINT, kr.REALM_ID, c.ID AS CLIENT_ID_JOINED
> FROM KEYCLOAK_ROLE kr
> LEFT JOIN `CLIENT` c ON kr.`CLIENT` = c.ID
> WHERE kr.`CLIENT` IS NOT NULL AND c.ID IS NULL

![Screenshot from 2024-11-18 12-24-13](https://github.com/user-attachments/assets/4401b424-c202-42aa-a3ca-993e45553d2d)

Deleting the COMPOSITE_ROLE entries and the entries from KEYCLOAK_ROLE + KC restart, made logging in the master console possible for me.

> SELECT *
> FROM COMPOSITE_ROLE
> WHERE COMPOSITE_ROLE.CHILD_ROLE IN (
> SELECT kr.ID
> FROM KEYCLOAK_ROLE kr
> LEFT JOIN `CLIENT` c ON kr.`CLIENT` = c.ID
> WHERE kr.`CLIENT` IS NOT NULL AND c.ID IS NULL
> )

Also I have to add that **due to a MISSING REALM_ID from the REALM** table, I had to delete entries from : CLIENT_SCOPE_CLIENT, CLIENT_SCOPE, CLIENT_SCOPE_ROLE_MAPPING, CLIENT_SCOPE_ATTRIBUTES, PROTOCOL_MAPPER, PROTOCOL_MAPPER_CONFIG. But **THAT DID NOT FIX MY ISSUE**. The KEYCLOAK_ROLE was the thing that broke the login.
